iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/mqtt/MQTTPublishHandler.java

```diff
@@ -0,0 +1,374 @@
+      schemas[i] = new MeasurementSchema(measurements.get(i), dataTypes[i]);
+    }
+    for (int i = message.getFields().size() + message.getTagKeys().size();
+        i
+            < message.getFields().size()
+                + message.getTagKeys().size()
+                + message.getAttributeKeys().size();
+        i++) {
+      dataTypes[i] = TSDataType.STRING;
+      columnCategories[i] = Tablet.ColumnCategory.ATTRIBUTE;
+      schemas[i] = new MeasurementSchema(measurements.get(i), dataTypes[i]);
+    }
+
+    final Tablet eventTablet =
+        new Tablet(
+            message.getTable(),
+            Arrays.asList(schemas),
+            Arrays.asList(columnCategories),
+            timestamps,
+            columns,
+            bitMaps,
+            1);
+
+    return new PipeRawTabletInsertionEvent(
+        true,
+        message.getDatabase().toLowerCase(),
+        null,
+        null,
+        eventTablet,
+        true,
+        pipeName,
+        creationTime,
+        pipeTaskMeta,
+        null,
+        false);
+  }
+
+  private void extractTree(final TreeMessage message, final MqttClientSession session) {
+    TSStatus tsStatus = null;
+    try {
+      TimestampPrecisionUtils.checkTimestampPrecision(message.getTimestamp());
+      tsStatus = checkAuthority(session);
+      if (tsStatus.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        LOGGER.warn(tsStatus.message);
+      } else {
+        final EnrichedEvent event = generateEvent(message);
+        if (!event.increaseReferenceCount(MQTTPublishHandler.class.getName())) {
+          LOGGER.warn(
+              "The reference count of the event {} cannot be increased, skipping it.", event);
+          return;
+        }
+        pendingQueue.waitedOffer(event);
+      }
+    } catch (Exception e) {
+      LOGGER.warn(
+          "meet error when polling mqtt source device {}, measurements {}, at time {}, because ",
+          message.getDevice(),
+          message.getMeasurements(),
+          message.getTimestamp(),
+          e);
+    }
+  }
+
+  private EnrichedEvent generateEvent(final TreeMessage message) throws QueryProcessException {
+    final String deviceId = message.getDevice();
+    final List<String> measurements = message.getMeasurements();
+    final long[] timestamps = new long[] {message.getTimestamp()};
+    final MeasurementSchema[] schemas = new MeasurementSchema[measurements.size()];
+    final TSDataType[] dataTypes =
+        message.getDataTypes() == null
+            ? new TSDataType[message.getMeasurements().size()]
+            : message.getDataTypes().toArray(new TSDataType[0]);
+    final List<String> values = message.getValues();
+    final BitMap[] bitMaps = new BitMap[schemas.length];
+    if (message.getDataTypes() == null) {
+      for (int index = 0; index < values.size(); ++index) {
+        dataTypes[index] = TypeInferenceUtils.getPredictedDataType(values.get(index), true);
```

Comment:
This may involve extra string compares for the original string value... May better predict it after the value is parsed